### ISSUE TITLE
Extremely sluggish behavior when in full screen on a 4K monitor

### ISSUE THREAD

**Issue by basavyr** [microsoft/terminal#3515]
# Environment

```none
Windows build number:
Platform      Version               VersionString
Win32NT     10.0.18362.0        Microsoft Windows NT 10.0.18362.0
Windows Terminal version: Version: 0.6.2951.0
```

# Steps to reproduce

Make the Windows Terminal go full-screen on a 4K monitor.
# Expected behavior
Fast and responsive (like it performs in non-full screen mode).

# Actual behavior
Super slow and sluggish behavior. There is at least a 2s delay for typing and using `ls` for example will result in a slow animation for listing the content in a directory. The Terminal is pretty much unusable if I go full-screen

I have a laptop with a 1080p display and I also have an external monitor connected via mini-display port. Monitor is 4K native 60 Hz, and it works without any issues. My laptop has also a dedicated GPU (nVidia GTX850m with the latest drivers). All other apps (i.e. Spotify, Microsoft Edge, Word etc) perform just fine when in full-screen on this 4K display. However, the Windows Terminal performs super slow when I go full screen on the monitor.
Terminal performs fine when I put it in full-screen on the laptop's main display.

**Comment by skyline75489** [comment]
Oh 4K display. That's probably where #778 come in handy. I've never tried Windows Terminal on 4K display yet.

**Comment by basavyr** [comment]
Thank you for responding to this :)
Regarding the #778 issue, I have some trouble understanding what do I have to change. I've only tried turning off acrylic, but that didn't help at all.

**Comment by miniksa** [comment]
Do you know how to get a WPR trace of it? @skyline75489 might be able to help you get one. It would be interesting to see if it's something we're doing or outside our control (and we can get it into the right team's hands.)

**Comment by basavyr** [comment]
@miniksa I got one WPR trace. Never tried it before, but I checked the documentation on Microsoft's website and I think I've managed to record the performance of this app. I hope it helps: I only had Windows Terminal as the running app on my desktop. I did put it in full screen mode and I start to type a bit inside the terminal, after that I just stopped the performance analyzer.

Trace file can be found here:

Please let me know if everything is ok!
Cheers :)

**Comment by ashao** [comment]
Just to an interesting system to this conversation as someone who's also using a 4K monitor hooked up to a laptop:
If I split the panes, the response noticeably improves. Splitting into 4 equal panes gives me the same (fast) response as being on my native laptop display.

**Comment by Jewsh-S** [comment]
I wanted to put a confirmation on this and I believe it is tied to [778](https://github.com/microsoft/terminal/issues/778).  This isn't just 4k, but general size of the app window.  If I make the app very very small, I get some what decent performance, but the bigger I make the app, the the worse the performance gets.  At 4k full screen, I get about 1.5 second of lag after I press a key until it actually outputs to the app window.  My first thought was that this might be my PC, because I normally force my laptop to disable the high performance GPU.  However, i force enabled my high performance GPU and the performance stayed the exact same.  This is definitely a drawing issue that makes the app almost unusable as the number of pixels that the terminal is using increases.

**Comment by miniksa** [comment]
@DHowett-MSFT, this is another one we're not going to be able to land.

**Comment by AH-Merii** [comment]
I still have the exact same issue after updating to the latest version 0.11.1121.0 . 

When running the terminal in full-screen on a 4k monitor, typing becomes sluggish. Sluggishness is directly proportional to window size, as window size increases so does the sluggishness.

Toggling acrylic off seems to improve the performance but not by much. Note: this is not an issue on when using the legacy cmd/powershell terminal.

**Comment by DHowett-MSFT** [comment]
Thanks for the corroborating report. This should be resolved when #5345 is in a released build.

**Comment by basavyr** [comment]
Awesome! Great job!

**Comment by AH-Merii** [comment]
> Thanks for the corroborating report. This should be resolved when #5345 is in a released build.

Hey, I just downloaded the new version, which was supposed to address this issue, I just wanted to report that I am still encountering the same problem regarding the sluggish typing experience. Unfortunately, all the same observations that I mentioned in my above comment still apply.

**Comment by DHowett-MSFT** [comment]
Reopening this one, but moving out of the 1.0 milestone. Sorry about that!

**Comment by DHowett-MSFT** [comment]
@basavyr does it work better for you?

**Comment by basavyr** [comment]
Hey there @DHowett-MSFT. Thank you for still continuing to work on this issue which I've encountered a while back.
Unfortunately, I can't provide any feedback, as for the last couple of months, I've switched entirely to _macOS_. As a result, I have no more Windows devices, so I am not able to test that.
I hope others can give feedback on this issue.

Best regards!

**Comment by skyline75489** [comment]
So I've finally got a 4K moniter and I still can not reproduce this. I've tried 100% and 200% scale. The performance is not very so bad to me. I wonder if there's something that's missing. Is it the graphic card of something? I'm using RX570 + Dell P2715Q.

**Comment by AH-Merii** [comment]
> So I've finally got a 4K moniter and I still can not reproduce this. I've tried 100% and 200% scale. The performance is not very so bad to me. I don't feel there's significant difference in performance between 4K and 1080p. I wonder if there's something that's missing. Is it the graphic card of something? I'm using RX570 + Dell P2715Q.

I have noticed a significant improvement after the update. However, the problem persists when activating acrylic mode. I don't think it's a problem with the graphics card because when I use powershell with decreased opacity it's performance remains the same.

TLDR: 
- Performance has significantly improved however it can still be improved further to be on par with CMD/Powershell. 
- Problem still persists when using Acrylic mode.

**Comment by DHowett** [comment]
Those of you who _are_ still seeing this problem **WITH ACRYLIC OFF**

What software are you using inside Terminal? Tmux? Screen? WSL? PowerShell? Anything?

**Comment by AH-Merii** [comment]
Sorry for not being specific enough but it's really odd. I tried using it with WSL, CMD, & Powershell, all their performance is comparable with each other, their performance is acceptable and much better than before.However, I am sure that performance slows down as I use it, I will try and test it extensively over the weekend, with acrylic off, so that I can reliably reproduce the issue.

**Comment by ashao** [comment]
The performance on my machine is great after the update to v1.0.1401.0. Both the native Powershell and Powershell within Windows Terminal (v1.0.1401.0) have the same responsiveness both with and without Acrylic.

This is on a Dell Precision M5510 with a Quadro M1000M.

**Comment by BaconTentacles** [comment]
> Those of you who _are_ still seeing this problem **WITH ACRYLIC OFF**
> 
> What software are you using inside Terminal? Tmux? Screen? WSL? PowerShell? Anything?

Anything and everything.  PowerShell 5, PowerShell 7, Git Bash, cmd.exe, you name it.  

Lenovo Legion Y740-15IRHg - nVidia GeForce RTX 2060 - 1080p laptop monitor.  Window is very slow to come up, and typing is absolutely painful.

**Comment by noomz** [comment]
I also have this issue with Acrylic turn on.

PC: Lenovo P1 gen 3
Graphics card: Intel integrated graphic card / Quadro T1000 max-q
Monitor: Internal display (4K, 225% zoom) / LG 27UL600-W (4K, 150% zoom)

Windows: 10 Pro 10.0.19042 Build 19042
GPU Driver: Lenovo NVIDIA Driver 27.21.14.5256 (14 Dec 2020)

Symptoms: When Window Terminal is on the external monitor, it works fine. But lag movement with the internal display. GPU 1 (Quadro) is used. The weird is that when the terminal window is halfway between 2 monitors (mainly on external, LG) it doesn't lag.

**Comment by JohnLukeBentley** [comment]
Extremly sluggish behaviour here too on a Desktop triple monitor system (with Windows terminal opened only on one monitor). Typing each letter of a command at the prompt has about a 0.5 to 2 sec delay. Resizing the Terminal (to be smaller) does not make it faster.

I'm on a 3 monitor setup (each only 1680x1050), in independent mode (without Nvidia Surround enabled). I haven't turned on Acrylic (as far as I know .. I'm unclear what Acrylic is ... but I don't have "Acrylic" anywhere in my Windows Terminal Settings file). Tried within Windows Terminal for: Powershell 7.1.0, Powerhsell 5.

(On my separate laptop system Windows Terminal works just fine).

WIndows Teriminal Version: 1.4.3243.0

Windows: 

    Edition:      Windows 10 Pro
    Version:      20H2
    OS build:     19042.685

Nvidia Drivers:

    451.67 (Game Ready Driver)

System:

    Operating System
      Windows 10 Pro 64-bit
    CPU
      Intel Core i7 3770 @ 3.40GHz  44°C
      Ivy Bridge 22nm Technology
    RAM
      8.00GB Dual-Channel DDR3 @ 788MHz (9-9-9-24)
    Motherboard
      Intel Corporation DZ77GA-70K (CPU 1)  32°C
    Graphics
      VX2262wm (1680x1050@59Hz)
      VX2262wm (1680x1050@59Hz)
      VX2262wm (1680x1050@59Hz)
      2047MB NVIDIA GeForce GTX 1660 Ti (ASUStek Computer Inc)  31°C
    Storage
      1863GB Western Digital WDC WD20NPVZ-00WFZT0 (SATA )  30°C
      953GB Samsung SSD 850 PRO 1TB (SATA (SSD))  34°C
    Optical Drives
      No optical disk drives detected
    Audio
      Realtek High Definition Audio

**Comment by JohnLukeBentley** [comment]
It's not clear if #7710 isn't the same issue.

**Comment by JohnLukeBentley** [comment]
I've removed my extreme sluggishness in WT by unintalling Powershell (core) 64bit, and installing Powershell (core) 32 bit. E.g. https://github.com/PowerShell/PowerShell/releases/download/v7.1.1/PowerShell-7.1.1-win-x86.msi

Can anyone else remove their sluggishness by doing the same?

**Comment by JohnLukeBentley** [comment]
Alas the update of Windows Terminal (automatically through the store) to 1.5.10271.0 has made the sluggishness return, with Powershell 32 bit installed.

**Comment by lhecker** [comment]
FYI: In my experience this issue is virtually fixed [by enabling the new AtlasEngine](https://devblogs.microsoft.com/commandline/windows-terminal-preview-1-13-release/#new-rendering-engine):
* Install Windows Terminal Preview 1.13
* Either
  * Select "Enable experimental text rendering engine" in the advanced settings of your default profile, or
  * Add `"experimental.useAtlasEngine": true` to your profile `"default"` object in your `settings.json`
* Restart Windows Terminal

On my 4K 120Hz display, selections, scrolling, drawing all occur at a stable, fluent 120 FPS and I couldn't observe any lags myself. I'd be extremely awesome if anyone else in this issue could confirm whether this new rendering engine still feels laggy or not. 🙂

**Comment by Fish1** [comment]
Switching to software rendering seemed to fix this issue for me.

**Comment by lhecker** [comment]
@Fish1 Do you have a Nvidia GPU?

**Comment by Fish1** [comment]
Nah AMD graphics.

I just noticed, windows did an auto update and replaced my latest driver with something they decided...

I am going to guess that is the issue, because yesterday I had the latest drivers from AMD (via AMD adrenaline) and everything was fine. I'll fix my drivers to verify.

**Comment by lhecker** [comment]
Do you perhaps use a FreeSync monitor? If so, could you try disabling dynamic refresh rate? Otherwise, the issue you're describing is definitely very very weird. Software rendering should definitely not be the only solution in this situation, because that is extremely taxing on the CPU. Unfortunately, I can't think of any way to debug this, apart from me debugging it on your exact device. 😟
You could check your debug logs with [DebugView](https://learn.microsoft.com/en-us/sysinternals/downloads/debugview), but I suspect there won't be any. If you know your way around debug tools, or just something like Process Explorer or System Informer, and notice anything weird with, let's say unusual CPU usage on any of the threads, please let me know!

**Comment by Fish1** [comment]
Yes, you are correct about the freesync.

I went to use the ScreenToGif tool to try and record my issue for you. But the moment I started recording, I managed to bluescreen windows (multiple times), simply by having both windows and ScreenToGif together.

But after disabling freesync (on my two monitors) everything seems to be working, and I can re-enable hardware rendering for windows terminal.

I don't really see anything happening in debug view.
